File: bin/dx_get_network_tests.pl
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#     http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# Copyright (c) 2016 by Delphix. All rights reserved.
#       
# Program Name : dx_get_network_tests.pl
# Description  : Get network test
# Author       : Marcin Przepiorowski
# Created      : 11 Aug 2016 (v2.0.0)
#
# 

use strict;
use warnings;
use JSON;
use Getopt::Long qw(:config no_ignore_case no_auto_abbrev); #avoids conflicts with ex host and help
use File::Basename;
use Pod::Usage;
use FindBin;
use Data::Dumper;

my $abspath = $FindBin::Bin;

use lib '../lib';
use Engine;
use Formater;
use Toolkit_helpers;
use Network_obj;
use Host_obj;
use Databases;

my $version = $Toolkit_helpers::version;

GetOptions(
  'help|?' => \(my $help),
  'd|engine=s' => \(my $dx_host),
  'remoteaddr=s' => \(my $remoteaddr),
  'type=s' => \(my $type),
  'last'   => \(my $last),
  'format=s' => \(my $format),
  'debug:i' => \(my $debug),
  'dever=s' => \(my $dever),
  'all' => (\my $all),
  'nohead' => \(my $nohead),
  'version' => \(my $print_version)
) or pod2usage(-verbose => 2, -output=>\*STDERR, -input=>\*DATA);

pod2usage(-verbose => 2, -output=>\*STDERR, -input=>\*DATA) && exit if $help;
die  "$version\n" if $print_version;

my $engine_obj = new Engine ($dever, $debug);
my $path = $FindBin::Bin;
my $config_file = $path . '/dxtools.conf';

$engine_obj->load_config($config_file);

if (defined($all) && defined($dx_host)) {
   print "Option all (-all) and engine (-d|engine) are mutually exclusive \n";
   pod2usage(-verbose => 2, -output=>\*STDERR, -input=>\*DATA);
   exit (1);
}

if ((defined($last) && (! defined($remoteaddr)))) {
   print "Option -last require remoteaddr to be defined \n";
   pod2usage(-verbose => 2, -output=>\*STDERR, -input=>\*DATA);
   exit (1);   
}

if (! defined($type)) {
   print "Option type is required \n";
   pod2usage(-verbose => 2, -output=>\*STDERR, -input=>\*DATA);
   exit (1);
}


my $output = new Formater();



if (lc $type eq 'latency') {
  $output->addHeader(
      {'name',                 35},
      {'remote host',          15},
      {'VDB found',            10}, 
      {'state',                15},
      {'average',              10},
      {'minimum',              10},
      {'maximum',              10},
      {'stddev',               10},
      {'count',                10},
      {'size',                 10},
      {'loss',                 10}
  );
} elsif (lc $type eq 'throughput') {
  $output->addHeader(
      {'name',                 35},
      {'remote host',          15}, 
      {'VDB found',            10}, 
      {'state',                15},
      {'direction',            15},
      {'no of conn',           10},
      {'throughput',           10},
      {'block size',           10}
  );
} elsif (lc $type eq 'dsp') {
  $output->addHeader(
      {'name',                 35},
      {'remote host',          15},  
      {'state',                15},
      {'direction',            15},
      {'no of conn',           10},
      {'throughput',           10},
      {'block size',           10}
  );
} else {
  print "Option type has unknown value - $type \n";
  pod2usage(-verbose => 2, -output=>\*STDERR, -input=>\*DATA);
  exit (1);
}  

# this array will have all engines to go through (if -d is specified it will be only one engine)
my $engine_list = Toolkit_helpers::get_engine_list($all, $dx_host, $engine_obj);

my $ret = 0;

for my $engine ( sort (@{$engine_list}) ) {
   # main loop for all work
   if ($engine_obj->dlpx_connect($engine)) {
    print "Can't connect to Dephix Engine $dx_host\n\n";
    next;
   };
  
   my $net  = new Network_obj ( $engine_obj, $debug );
   my $hosts = new Host_obj    ( $engine_obj, $debug );
   my $databases = new Databases ( $engine_obj, $debug);
   
   my $testlist;

   if (defined($remoteaddr)) {
      my @templist;
      my @hostlist;
      if (lc $remoteaddr eq 'all') {
        @hostlist = $hosts->getAllHosts();
      } else {
        for my $h (split (',', $remoteaddr)) {
          my $hostref = $hosts->getHostByAddr($h);
          if (!defined($hostref)) {
             print "Remote host with addr $h not found in Delphix Engine\n";
             $ret=$ret+1;
             next;
          }
          push(@hostlist,$hostref);          
        }
      }
      
      
      #my @hostlist 
      for my $hostitem (sort @hostlist) {
         my $testref;
         if (defined($last)) {
            if (lc $type eq 'latency') {
              $testref = $net->getLatencyLastTests($hostitem);
              if (defined($testref)) {
                push(@templist, @{$testref});
              }
            } elsif (lc $type eq 'throughput') {
               $testref = $net->getThroughputLastTests($hostitem);
               if (defined($testref)) {
                 push(@templist, @{$testref});
               }
            } else {
               $testref = $net->getDSPLastTests($hostitem);
               if (defined($testref)) {
                 push(@templist, @{$testref});
               }
            }
         } else {
            if (lc $type eq 'latency') {
               $testref = $net->getLatencyTestsList($hostitem);
               if (defined($testref)) {
                 push(@templist, @{$testref});
               }
            } elsif (lc $type eq 'throughput') {
               $testref = $net->getThroughputLastTests($hostitem);
               if (defined($testref)) {
                 push(@templist, @{$testref});
               }
            } else {
               $testref = $net->getDSPTestsList($hostitem);
               if (defined($testref)) {
                 push(@templist, @{$testref});
               }
            }
         }
      }
      $testlist = \@templist;
      
      
      
   } else {
      if (lc $type eq 'latency') {
         $testlist = $net->getLatencyTestsList();
      } elsif (lc $type eq 'throughput') {
         $testlist = $net->getThroughputTestsList();
      } else {
         $testlist = $net->getDSPTestsList();
      }
   }
   
   for my $netitem (@{$testlist}) {
     

      my $hostname;
      my $hostref = $net->getHost($netitem);

      if (defined($hostref)) {
         $hostname = $hosts->getHost( $hostref )->{name};
      } else {
         $hostname = 'N/A';
      }
      
      my @dblist = $databases->getDBForHost($hostname);
      
      my $dbtype = grep { ($databases->getDB($_))->getType() eq 'VDB' } @dblist ; 
      my $dbtype_disp = $dbtype > 0 ? 'YES' : 'NO'; 
      
      if (lc $type eq 'latency') {
         $output->addLine(
          $net->getName($netitem),
          $hostname,
          $dbtype_disp,
          $net->getState($netitem),
          $net->getLatencyAvg($netitem),
          $net->getLatencyMin($netitem),
          $net->getLatencyMax($netitem),
          $net->getLatencyStdDev($netitem),
          $net->getLatencyCount($netitem),
          $net->getLatencySize($netitem),
          $net->getLatencyLoss($netitem)
         );
      } elsif (lc $type eq 'throughput') {
         $output->addLine(
           $net->getName($netitem),
           $hostname,
           $dbtype_disp,
           $net->getState($netitem),
           $net->getTestDirection($netitem),
           $net->getTestConnections($netitem),
           $net->getTestRate($netitem),
           $net->getTestBlockSize($netitem)
         );
      } else  {
         $output->addLine(
           $net->getName($netitem),
           $hostname,
           $net->getState($netitem),
           $net->getTestDirection($netitem),
           $net->getTestConnections($netitem),
           $net->getTestRate($netitem),
           $net->getTestBlockSize($netitem)
         );
      }
   }

}

Toolkit_helpers::print_output($output, $format, $nohead);

exit $ret;


__DATA__

=head1 SYNOPSIS

 dx_get_network_tests.pl [ -engine|d <delphix identifier> | -all ] -type latency|throughput
                         [ -remoteaddr env_ip ] 
                         [ -last]  
                         [ -format csv|json ]  
                         [ -help|? ] [ -debug ]

=head1 DESCRIPTION

Get the results of the network tests

=head1 ARGUMENTS

Delphix Engine selection - if not specified a default host(s) from dxtools.conf will be used.

=over 10

=item B<-engine|d>
Specify Delphix Engine name from dxtools.conf file

=item B<-all>
Display databases on all Delphix appliance

=item B<-type latency|throughput>
select a type of test to display

=back

=head1 OPTIONS

=over 3

=item B<-remoteaddr env_ip>
Filter results to a env_ip

=item B<-last>
List only last results of test for a specified env_ip


=item B<-format>
Display output in csv or json format
If not specified pretty formatting is used.

=item B<-help>
Print this screen

=item B<-debug>
Turn on debugging

=back




=cut

File: bin/dx_get_network_tests.py
#!/usr/bin/env python3
"""Python port of dx_get_network_tests.pl (lightweight CSV/JSON output)
"""
import argparse
import sys
import os

ROOT = os.path.dirname(os.path.dirname(__file__))
sys.path.insert(0, os.path.join(ROOT, 'lib', 'py'))

# Prefer mock engine when explicitly requested or if real engine import fails
try:
    from engine import Engine as _RealEngine
except Exception:
    _RealEngine = None
try:
    import mock_engine as _mock
except Exception:
    _mock = None
Engine = _mock.MockEngine if (os.environ.get('DXTOOLKIT_USE_MOCK') == '1' and _mock) else (_RealEngine or (_mock.MockEngine if _mock else None))
from network_obj import NetworkObj
from host_obj import HostObj
from databases import Databases
from formater import Formater
import toolkit_helpers


def parse_args(argv):
    p = argparse.ArgumentParser(description="Get network test results")
    p.add_argument('-d', '--engine', dest='dx_host')
    p.add_argument('-all', action='store_true')
    p.add_argument('-configfile', '-c', dest='config_file')
    p.add_argument('-type', dest='test_type', choices=['latency', 'throughput', 'dsp'], required=True)
    p.add_argument('-remoteaddr', dest='remoteaddr')
    p.add_argument('-last', action='store_true')
    p.add_argument('-format', dest='fmt', choices=['csv', 'json'])
    p.add_argument('-nohead', action='store_true')
    p.add_argument('-debug', dest='debug', type=int, nargs='?', const=1)
    p.add_argument('-dever', dest='dever')
    p.add_argument('-version', action='store_true')
    return p.parse_args(argv)


def main(argv):
    args = parse_args(argv)
    if args.version:
        print(toolkit_helpers.version)
        return 0

    if not toolkit_helpers.ensure_config_file(args.config_file):
        return 1

    if args.all and args.dx_host:
        print('Option all (-all) and engine (-d|engine) are mutually exclusive')
        return 1

    if args.last and (not args.remoteaddr):
        print('Option -last requires -remoteaddr to be defined')
        return 1

    eng = Engine(args.dever, args.debug)
    try:
        eng.load_config(args.config_file)
    except FileNotFoundError:
        print(f"ERROR: config file not found: {args.config_file}")
        return 1
    except Exception as exc:
        print(f"ERROR: failed to load config file {args.config_file}: {exc}")
        return 1

    engine_list = toolkit_helpers.get_engine_list(args.all, args.dx_host, eng)
    ret = 0

    for engine_name in sorted(engine_list):
        if eng.dlpx_connect(engine_name):
            print(f"Can't connect to Dephix Engine {engine_name}")
            ret += 1
            continue

        net = NetworkObj(eng, args.debug)
        hosts = HostObj(eng, args.debug)
        dbs = Databases(eng, args.debug)

        # Prepare output headers
        out = Formater()
        if args.test_type == 'latency':
            out.addHeader({'engine': 20}, {'name': 35}, {'remote host': 20}, {'VDB found': 10},
                          {'state': 15}, {'average': 10}, {'minimum': 10}, {'maximum': 10},
                          {'stddev': 10}, {'count': 10}, {'size': 10}, {'loss': 10})
        else:
            out.addHeader({'engine': 20}, {'name': 35}, {'remote host': 20}, {'VDB found': 10},
                          {'state': 15}, {'direction': 12}, {'no of conn': 12}, {'throughput': 12}, {'block size': 12})

        # Decide hostrefs
        hostrefs = []
        if args.remoteaddr:
            if args.remoteaddr.lower() == 'all':
                hostrefs = hosts.getAllHosts()
                if not hostrefs:
                    hostrefs = [None]
            else:
                for h in args.remoteaddr.split(','):
                    href = hosts.getHostByAddr(h)
                    if href:
                        hostrefs.append(href)
                    else:
                        print(f"Remote host with addr {h} not found in Delphix Engine")
                        ret += 1
        else:
            hostrefs = [None]

        # Collect tests
        testrefs = []
        for href in hostrefs:
            if args.last:
                if args.test_type == 'latency':
                    testrefs.extend(net.getLatencyLastTests(href))
                elif args.test_type == 'throughput':
                    testrefs.extend(net.getThroughputLastTests(href))
                else:
                    testrefs.extend(net.getDSPLastTests(href))
            else:
                if args.test_type == 'latency':
                    testrefs.extend(net.getLatencyTestsList(href))
                elif args.test_type == 'throughput':
                    testrefs.extend(net.getThroughputTestsList(href))
                else:
                    testrefs.extend(net.getDSPTestsList(href))

        # If filtering by engine hosts returned nothing, fall back to all tests
        if (not testrefs) and args.remoteaddr and args.remoteaddr.lower() == 'all':
            if args.test_type == 'latency':
                testrefs.extend(net.getLatencyTestsList())
            elif args.test_type == 'throughput':
                testrefs.extend(net.getThroughputTestsList())
            else:
                testrefs.extend(net.getDSPTestsList())
            if testrefs:
                print("No network tests matched engine hosts; falling back to all available tests.")

        # Render lines
        for tref in testrefs:
            hostref = net.getHost(tref)
            hostobj = hosts.getHost(hostref) if hostref else {'name': 'N/A'}
            hostname = hostobj.get('name', 'N/A')
            vdb_list = dbs.getDBForHost(hostname)
            vdb_found = 'YES' if any(dbs.getDB(t) and dbs.getDB(t).get('provisionContainer') for t in vdb_list) else 'NO'
            state = net.getState(tref)
            name = net.getName(tref)

            if args.test_type == 'latency':
                out.addLine(engine_name, name, hostname, vdb_found, state,
                            net.getLatencyAvg(tref), net.getLatencyMin(tref), net.getLatencyMax(tref),
                            net.getLatencyStdDev(tref), net.getLatencyCount(tref), net.getLatencySize(tref), net.getLatencyLoss(tref))
            else:
                out.addLine(engine_name, name, hostname, vdb_found, state,
                            net.getTestDirection(tref), net.getTestConnections(tref), net.getTestRate(tref), net.getTestBlockSize(tref))

        toolkit_helpers.print_output(out, args.fmt or 'csv', args.nohead)

    return ret


if __name__ == '__main__':
    sys.exit(main(sys.argv[1:]))
